Custom\manual waveguide modes definition for Excitations

gadiLhv opened the pull request:
As I am getting the hang of the Git protocol, I can start this, seperately. This was (re)based around the original repository master, rather than my one, which has the absorbers. In the future I'll make a habit of opening branches for new features.

I'll go over the remarks from the previous PR

> The python interface is wrong? It talks about coors? (coordinates??) and "Weights" but only has the latter?

The line that guided me here was "make as little damage as possible". This was for two main reasons:
1. Avoid code repetitions, where possible.
2. Backwards compatibility with existing code.

Hence, the fields for the mode consist of three columns of coordinates, and three columns of field values: [X,Y,Z,Ex,Ey,Ez]. Replace with H at your own convenience.

> The mode should be purely 2D on a plane right? Why does it accept coords in x,y and z and why weights in x/y/z? I would expect only x/y-coordinates (or maybe more general x0 and x1) and a 2D array of weights? That does all make little sense to me?

Three reasons, this time
1. Make as little damage as possible to the original code. This is clearer in openEMS. I know they are seperate, but still. I do a "nearest neighbor" search later, and it makes a lot more sense.
2. The mode functions are also given in all three axes, so it made sense to me.
3. This may sound arbitrary, but it isn't. When I output the geometry data to my mode solver (I am hoping to finalize that some day with the new library I'm using, so I can share it), I am simply snipping it out of the 3D structure. In case the normal is z, it's simple enough. But if the normal is y, I need to rotate the geometry 90 degrees...
To avoid all this mess, I simply extract all three coordinates. I'll give this some more re-thinking.

> I really do not like to give this class possibly large arrays of data like this and none of these values are stored and read into xml? This would be needed if at some point the Matlab/Octave interface want to use this feature too.

I know. I didn't find any built-in XML functions for handling arrays, except for the mesh processing. Care to give me a hint?

> I really do not like that this is implemented twice in the excitation and the probe! (Large) redundant code is a no-no!!

I'll think how to implement these functions just one time. Missed this remark while re-basing the code.

>Create a new file format (maybe plain csv or json or hdf5) that holds all required info about the modes, e.g. x and y grid coordinates and the mode on a 2D plane
The excitation and probes use a function SetWeightFromFile with some additional info's (see next point)
Supply an additional translation (and maybe rotation) information to allow to reuse the mode file for multiple ports at multiple locations and rotations

This will take some figuring out. CSV is my weapon of choice, I guess. Easy to manipulate both in M-script and python. If I make this a 2D file, how do I handle the coordinate shift? Rotation is the easy bit, but where is (0,0)?

> This approach would allow openEMS to have one new (utility) method to read and interpolate this custom mode for both the excitation and the probe and thus avoid redundant code and memory usage

gadiLhv committed:
Initial commit to start bringing this feature up to speed.
Not compiling, yet.

gadiLhv committed:
Commit ready to begin PR

Commit:
Compiles, no test sript

gadiLhv commented:
Hey @thliebig 

Took me long enough, but I have a parser for CSV files with E and H data. 
Is it best to put it in CSXCAD or OpenEMS?

gadiLhv committed:
Not working yet - First commit with mode parser

gadiLhv committed:
Add mode parser files

gadiLhv committed:
Push attempt with EGit

gadiLhv committed:
Resolved direction issue, still not finished.

Didn't resolve the local coordinates shift, yet.
Didn't figure out how to write in XML the mode file name.

gadiLhv committed:
Ready for first compilation attempt.

gadiLhv committed:
Fix CSPropExcitation

gadiLhv committed:
Compiles, untested

1. Minor C++ fixes
2. Fixed python bindings
3. Fixed CMakeLists.txt

gadiLhv committed:
Added functionality to CSProbeBox

gadiLhv committed:
Added XML functionality and syntax fixes.

gadiLhv committed:
Added XML functionality

Syntax fixes

gadiLhv committed:
Compiles with openEMS, untested!

gadiLhv committed:
Switch moduld to module

gadiLhv committed:
Minor fix to CSPropExcitation

gadiLhv committed:
Debugged ModeFileParse

1. Bilinear interpolation errors
2. Parsing error - line too short.

gadiLhv commented:
Hey @thliebig 

Finally got a hold of this, or at least I think. Rebased, tested, the works.
I'm guessing you will have some corrections and questions.
Also, a similar post will be in the openEMS counterpart of this.

Cheers

gadiLhv committed:
Removed stdlib import

No longer necessary as there are no more c-style memory allocations in the python interface.

gadiLhv committed:
Resolved bug with ModeFileParser.cpp transpose

thliebig reviewed (commented):
Additionally there are many random blank lines in odd places?

gadiLhv committed:
PR comment resolution

gadiLhv committed:
Added tests for interpolant

gadiLhv commented:
I did that. It compiled and everything, but when I run the unit test it doesn't see my additions.
Also, there is a segmentation fault on test_dump.
Really don't know how to proceed. 
When I try to use plain update_openEMS.sh, I'm getting some error about Cythom module missing. It's not. 

Let me know what to do. I'm really on the verge of ditching this.

gadiLhv committed:
Final fixes. Can't install with new workflow

thliebig commented:
Well I was able to checkout, rebase and build it just fine. Not sure whats up with your setup (or mine). The tests obviously can not work like this yet and they are already a bit deeper than I intended. 
Maybe you just remove the last commit, I will do my final review and than we call it good and I will maybe have a look at this unittest later. My intention was more to test the GetWeightedExcitation C++-function, but I realize that it does not have a python interface to begin with...

gadiLhv commented:
> > I can port GetWeightedExcitation to python, but I can't test. It's pretty much the same though...
> 
> That does make little sense, just "unpush"? The last commit and we go from there...

I edited a second later.

This is the line it uses:
 ```cpp
m_ModeFile.linInterp2(loc_coords[nPyp],loc_coords[nPypp],Wny)
```
With this test, you can use both interpolants.

tmpusr123 commented:
Hi @thliebig, happy new year — did you have a chance to build/test this locally, or are there any more blockers to merging?

thliebig commented:
Sorry but the test are still failing and I tried to fix some obvious issues, but the tests still fail...
Please get this fixed and get the install workflow figures out? 
Ideally also rebase on the current master and force push an update too

gadiLhv commented:
Oh well
I'm hoping them LLMs will help me wrestle with this. I'll get to this probably next week. 
Appreciate the effort, @thliebig. Will finish this soon enough.

gadiLhv committed:
Fixed CSProperties unit test

1) Fixed the unit test. Added relevant assertions
2) Found bug in ModeFileParser.cpp - Nearest neighbor case

gadiLhv committed:
Parentheses error

gadiLhv commented:
It wasn't a simple rebase, but with the help of friendly robots, I managed.

The results for this
`python -m unittest test_CSProperties.py`
is this
```
.............
----------------------------------------------------------------------
Ran 13 tests in 0.042s
```

I verified that it works with openEMS, as well, although that's irrelevant here.
Let me know if I missed something, @thliebig.

Cheers

gadiLhv committed:
Removed duplicate entry in CSProperties.pxd

gadiLhv commented:
I noticed that the linux tests fail because of lack of scipy. What should I do?

thliebig commented:
We should avoid the scipy dependency (only for this). You can just check against fixed numbers instead of using the interpolator? Usually putting fixed numbers in code is a no-no. But in this tests it is okay as we expect the result to always be the exact same number no matter what. 
Maybe comment out the generating lines thus in the future new numbers can be created and their "correctness" checked.

gadiLhv commented:
I checked. Numpy has linear interp built in. I'll write NN manually.

Will finish this by tomorrow.

gadiLhv committed:
Removed scipy dependence

test_CSProperties.py was embedded with linear and nearest-neighbor interpolation schemes.
The test is now with more points.

gadiLhv commented:
Green lights all around, @thliebig

gadiLhv commented:
To try and not mix further the already mixed salad here.

@biergaizi, @thliebig, if the only issue to make this work with openEMS better is to rename the mode parser as CSModeFileParser and add the `CSXCAD_EXPORT` predef, say the word. 
I'll expose as public as little as possible, per your suggestion.
I saw there Re additional exclusions for this logic, e.g. `ParameterCoord`. 

Let me know

gadiLhv commented:
@biergaizi, I reviewed the code. Everything that is exposed is necessary. 

@thliebig, let me know what is the best course of action here. Simply add `CSXCAD_EXPORT` or rename the class and files first.

thliebig commented:
Why would you need to rename something? It just sounds like you need to add some "CSXCAD_EXPORT" at some places? 
But I need to have a closer look at the Windows error message I guess

thliebig commented:
As far as I see it the class "ModeFileParser" is not exposed. Just add a "CSXCAD_EXPORT" in front of it and all should be good.... maybe :smirk:

gadiLhv committed:
Allow CSXCAD_EXPORT

To resolve the openEMS MSVC compilation issues, CSXCAD_EXPORT was added
to ModeFileParser.h

gadiLhv commented:
Can do! 
If you can, please trigger a check in openEMS as well, so we know if this was enough.

thliebig commented:
Seems good now

biergaizi commented:
> @thliebig: As far as I see it the class "ModeFileParser" is not exposed. Just add a "CSXCAD_EXPORT" in front of it and all should be good.... maybe 

Please reconsider, and address my comment in https://github.com/thliebig/openEMS/pull/175#issuecomment-3797022001. Everything in ModeFileParser will permanently become CSXCAD's public API, and in theory anyone who uses CSXCAD will be able to link against it and use it. But this class does not follow CSXCAD's naming conventions, and it will be too late to change it if it's accepted as a stable API.

thliebig commented:
@biergaizi, sorry I somehow missed that comment completely. I agree with your summary there but it exposes more like a different issue. We need the "CSXCAD_EXPORT" to make (the current implementation) equal to whats available for Linux and fix the current issue. So it needs to be there. But we should address the points you made too.
I think we should go with points 1+2 for now by making it a better API. There is not really a way to truly make an API "internal" in C++ right?

But another point I would like to look at is, why does openEMS need to know all this at all? Ideally openEMS just wants the mode profile and does not want to know how and where that came from (function vs file)? Not sure if that is feasible or not. But if we could make that unnecessary, we could make the ModeFileParser truly "internal" maybe....

But it does not hurt anything if @gadiLhv could look at and fix this to the CSXCAD naming convention (fully internal or not), thanks.

gadiLhv commented:
I'll address what's relevant to me, if you may.

> But another point I would like to look at is, why does openEMS need to know all this at all?

I'll start with the complicated part. The mode port factually has 2 parts.
1. Excitation - This part is 100% internal in CSXCAD. High fives all around. The exception is the two openings I left in `CSPropExcitation`, for the unit tests.
2. Field probe + mode match integral. This is where it gets messy. There is no internal functionality available here. If you look in openEMS (@theliebig master branch), there is direct access to CSFunctionParser [here](https://github.com/thliebig/openEMS/blob/32c5c6b537b33a8b70f9ba4f5c9a8ecbb12777b3/Common/processmodematch.cpp#L123), so I followed suit. The other option requires more significant changes than the intentions of this PR.

>  But it does not hurt anything if @gadiLhv could look at and fix this to the CSXCAD naming convention (fully internal or not), thanks.

Except for renaming it to something like CSModeParser, anything else?

thliebig commented:
Well name of the files, name of the class and name of the methods. As well as maybe checking which methods need to be public and which one could be made private.

gadiLhv committed:
Changed ModeFileParser Naming Convention

According to latest PR request:
1) Changed the ModeFileParser.cpp and ModeFileParser.h naming convention
2) Exposed as little as possible from the CSModeFileParser class methods
as public.

No python interface added to CSModeFileParser

gadiLhv commented:
> Well name of the files, name of the class and name of the methods. As well as maybe checking which methods need to be public and which one could be made private.

Applied this, to the best of my understanding.

gadiLhv commented:
Hey @thliebig, 
Did you have a chance to test this? I'd like to know if there are any other changes necessary.

thliebig commented:
No not yet but I am hopeful that we have it now. If you really want, you can maybe have a look if you can add this to the Matlab/Octave interface? Should not be to difficult?

gadiLhv commented:
This is actually more relevant for the openEMS PR, I think. 

In any account, it's a different demon I need to tackle, as it's 100% XML stuff.
Please treat this as non blocking. Should I get to it in the near week, Ok. If not, I boy-scout promise to open a new PR specifically for this.

gadiLhv committed:
Initial Commit

Still need to set the atributes or whateveer for SetExcitationFile.m

gadiLhv committed:
Added support for mode file names

Not working, yet

gadiLhv committed:
XML read bug fix

1. Fixed bug in CSPropExcitation.cpp. Erroneously flagged mode file read
as true.
2. Added the PropDir property when mode files are read

gadiLhv committed:
Minor fixes

Fixed NormDir XML print logic

gadiLhv commented:
Hello again

I did find a critical bug in the XML reader due to adding the Matlab functionality, so it's fixed.

Matlab/Octave code in 2nd PR.

thliebig commented on src/CSPropProbeBox.cpp:
```diff
@@ -100,10 +100,9 @@ bool CSPropProbeBox::Write2XML(TiXmlNode& root, bool parameterised, bool sparse)
 
 	prop->SetAttribute("Number",(int)uiNumber);
 	if ((m_NormDir>0) && (m_NormDir<3))
```
Should this not be >=0 ?

gadiLhv replied:
I have no idea.
I'm not using this attribute, and I didn't even write these lines :)

https://github.com/thliebig/CSXCAD/blob/f5e47643a28d6efd42cc10b61b848903e6599581/src/CSPropProbeBox.cpp#L59-L60

gadiLhv replied:
The only change I did here was remove a tab...

thliebig replied:
Hehe, yes of course. I will have a closer look. This just looks wrong, but maybe not... But indeed, not related to this PR in any case

gadiLhv replied:
Ok then. Can I mark this resolved?
